python/ray/tune/impl/tuner_internal.py

```diff
@@ -123,6 +126,55 @@ def __init__(
         trainable_ckpt = os.path.join(self._experiment_checkpoint_dir, _TRAINABLE_PKL)
         with open(trainable_ckpt, "wb") as fp:
             pickle.dump(self._trainable, fp)
+        self._maybe_warn_resource_contention()
+
+    def _expected_utilization(self, cpus_per_trial, cpus_total):
+        num_samples = self._tune_config.num_samples
+        if num_samples < 0:  # TODO: simplify this in Tune
+            num_samples = math.inf
+        concurrent_trials = self._tune_config.max_concurrent_trials or 0
+        if concurrent_trials < 1:  # TODO: simplify this in Tune
+            concurrent_trials = math.inf
+
+        actual_concurrency = min(
+            (cpus_total // cpus_per_trial, num_samples, concurrent_trials)
+        )
+        return (actual_concurrency * cpus_per_trial) / cpus_total
+
+    def _maybe_warn_resource_contention(self):
+        trainable = self._convert_trainable(self._trainable)
+
+        # This may not be precise, but we don't have a great way of
+        # accessing the scaling config.
+        scaling_config = None
+        get_scaling_config = getattr(trainable, "base_scaling_config", None)
+        if callable(get_scaling_config):
+            scaling_config = get_scaling_config()
+
+        if not isinstance(scaling_config, ScalingConfig) or getattr(
+            scaling_config, "_max_cpu_fraction_per_node"
+        ):
+            return
+
+        has_dataset = getattr(trainable, "has_dataset", False)
+
+        cpus_per_trial = scaling_config.total_resources.get("CPU", 0)
+        cpus_left = ray.available_resources().get("CPU", 0) + 0.001  # avoid div by 0
+        # TODO(amogkam): Remove this warning after _max_cpu_fraction_per_node is no
+        #  longer experimental.
+        if has_dataset and self._expected_utilization(cpus_per_trial, cpus_left) > 0.8:
+            warnings.warn(
+                "Executing `.fit()` may leave less than 20% of CPUs in "
+                "this cluster for Dataset execution, which can lead to "
+                "resource contention or hangs. To avoid this, explicitly "
+                "reserve at least 20% of node CPUs "
+                "for Dataset execution by "
+                "setting _max_cpu_fraction_per_node = 0.8 in the Trainer "
+                "scaling_config. Not doing so can  See "
+                "https://docs.ray.io/en/master/data/key-concepts.html"
+                "#example-datasets-in-tune for more info.",
```

Comment:
This has moved to https://docs.ray.io/en/master/data/dataset-internals.html#datasets-and-tune